platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,437 @@
+     */
+    HashObject active_roster_hash = 7;
+}
+```
+
+It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.
+This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, and an
+optional current Candidate Roster. Insertion of rosters will be controlled by ensuring that the acceptance of a new
+Candidate Roster invalidates and removes the current candidate roster. Therefore, the map is expected to have a maximum
+size of three elements.
+There will be new fields in PlatformState - `candidateRosterHash` and `activeRosterHash` - such that at adoption time,
+the app will set the roster through the API and the platform code inserting this roster into the roster map, alongside
+setting the `candidateRosterHash` field in the PlatformState. If a `candidateRosterHash` hash entry already exist in the
+map of Rosters, it will be discarded. That is, setting a candidate roster is an idempotent operation.
+The `activeRosterHash` will be private to the platform and will not be settable or modifiable from outside the platform.
+
+This indirection (using a Map instead of a Singleton) avoids moving data around in the merkle state which requires data
+to be copied.
+When a map is updated, only the key-value pair that changes needs to be added to the block stream.
+Another benefit of this approach is that adoption trigger becomes simple (app sets the roster) with delineated
+responsibilities between the app and the platform.
+
+
+###### Roster Validity
+
+In simple terms, the following constitutes a valid roster:
+
+1. The roster must have at least one RosterEntry.
+2. At least one RosterEntry/ies must have a non-zero weight.
+3. All RosterEntry/ies must have a valid X509Certificate.
+4. All RosterEntry/ies must have a valid PairingPublicKey.
+5. All RosterEntry/ies must have at least one gossip Endpoint.
+6. The RosterEntry/ies must be specified in order of ascending node id.
+7. All ServiceEndpoint/s must have a valid IP address or domain name (mutually exclusive), and port.
+8. The roster must have a unique NodeId for each RosterEntry.
+
+Note that a roster can be valid, but not accepted by the platform.
+For example, if a new candidate roster is set via the API, but its hash evaluates to the hash of the existing
+candidate roster, the new roster will be discarded. The operation has no effect.
+
+### Core Behaviors, in summary
+
+- Roster Creation: App will create a Candidate Roster from the Candidate Address Book (CAB).
+- Roster Submission: App will trigger roster submission by setting the `candidate_roster_hash` field in the
+  PlatformState.
+- Roster Adoption: The platform will vote to adopt the last submitted Candidate Roster when it is ready, which will be
+  on software upgrade.
+- Roster Replacement: If a new candidate roster is submitted before the previous one is adopted, the corresponding new
+  Candidate Roster will replace the old one.
```

Comment:
How does the App learn about the new Roster?

Reply:
The app will create a new Roster from the address book whenever it decides it's time to start working on adopting a new Roster.

Reply:
@netopyr   The roster replaces the AddressBook in the `ConsensusRound` when it is handled.  It can detect that it is a new consensus roster when it gets back the candidate roster it set.  https://github.com/hashgraph/hedera-services/blob/develop/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/ConsensusRound.java

If it would make life easier, we could find a more direct way of communication of the change.

Reply:
Does this mean that `ConsensusRound` will become part of the public API?

I like the idea of getting the full roster because we do not need to keep track of potentially multiple candidate rosters. In my opinion, we do not need a more direct way to track the changes. We can copy all values into the address book if we detect a new active roster.

How can we quickly determine if the active roster changed? Comparing each field for each node in each new round would be wasteful.

Reply:
The public API state change that reflects adoption of the candidate roster is in the `RosterState` singleton.    The value of the candidate roster hash will move to the  active roster position, at index 0 in the `round_id_to_roster_pair`.   The round value in the pair indicates the round that the roster became active. 

```
message RosterState {
    byte candidate_roster_hash = 1;
    repeated RoundRosterPair round_id_to_roster_pair = 2;
}
```